spec/validators/waste_exemptions_engine/main_person_validator_spec.rb

```diff
@@ -30,188 +30,74 @@ module WasteExemptionsEngine
 
-      %i[limited_company limited_liability_partnership local_authority charity].each do |business_type|
         context "when the business type is #{business_type}" do
           let(:transient_registration) { create(:transient_registration, business_type) }
           subject(:validatable) { Test::MainPersonValidatable.new(transient_registration) }
 
-          context "and the number of people is two" do
-            before(:each) do
-              transient_registration.transient_people = [build(:transient_person), build(:transient_person)]
-            end
-
-            it "confirms the object is valid" do
-              expect(validatable).to be_valid
-            end
-
-            it "the errors are empty" do
-              validatable.valid?
-              expect(validatable.errors).to be_empty
-            end
-          end
-
-          context "and the number of people is one" do
-            before(:each) { transient_registration.transient_people = [build(:transient_person)] }
-
-            it "confirms the object is valid" do
-              expect(validatable).to be_valid
-            end
-
-            it "the errors are empty" do
-              validatable.valid?
-              expect(validatable.errors).to be_empty
-            end
-          end
-
-          context "and there aren't any people" do
-            let(:error_message) do
-              Helpers::Translator.state_error_message(validatable, :not_enough_main_people, :one)
-            end
-
-            before(:each) { transient_registration.transient_people = [] }
-
-            it "confirms the object is invalid" do
-              expect(validatable).to_not be_valid
-            end
-
-            it "adds a single validation error to the record" do
-              validatable.valid?
-              expect(validatable.errors[:base].count).to eq(1)
-            end
-
-            it "adds an appropriate validation error" do
-              validatable.valid?
-              expect(error_message).to_not include("translation missing:")
-              expect(validatable.errors[:base]).to eq([error_message])
+          min_people = allowed_num_people.rassoc(true).first # This will get the first key that has a value of true.
+
+          allowed_num_people.each do |num_people, is_allowed|
+
+            context "and the number of people is #{num_people}" do
+              before(:each) do
+                people = []
+                num_people.times { people << build(:transient_person) }
+                transient_registration.transient_people = people
+              end
+
+              if is_allowed
+                it "confirms the object is valid" do
+                  expect(validatable).to be_valid
+                end
+
+                it "the errors are empty" do
+                  validatable.valid?
+                  expect(validatable.errors).to be_empty
+                end
+              else
+                let(:error_message) do
```

Comment:
What do you think about moving the logic for determining the error message into a separate method(s)? I think iterating over `expectations` is definitely the way to go, but the result is a little complex, and I think hiding that complexity could improve the overall readability.

Reply:
Fixed in d90ead8.